pkg/storegateway/batch_series.go

```diff
@@ -0,0 +1,846 @@
+	} else {
+		b.stats.update(func(stats *queryStats) {
+			stats.seriesHashCacheHits++
+		})
+	}
+	return hash
+}
+
+func shardOwned(shard *sharding.ShardSelector, hasher seriesHasher, id storage.SeriesRef, lset labels.Labels) bool {
+	if shard == nil {
+		return true
+	}
+	hash := hasher.Hash(id, lset)
+
+	return hash%shard.ShardCount == shard.ShardIndex
+}
+
+func (s *BucketStore) batchSetsForBlocks(ctx context.Context, req *storepb.SeriesRequest, blocks []*bucketBlock, chunkReaders *chunkReaders, shardSelector *sharding.ShardSelector, matchers []*labels.Matcher, chunksLimiter ChunksLimiter, seriesLimiter SeriesLimiter) (storepb.SeriesSet, *hintspb.SeriesResponseHints, *queryStats, func(), error) {
+	resHints := &hintspb.SeriesResponseHints{}
+	mtx := sync.Mutex{}
+	batches := make([]unloadedBatchSet, 0, len(blocks))
+	g, ctx := errgroup.WithContext(ctx)
+	stats := &queryStats{}
+	cleanups := make([]func(), 0, len(blocks))
+
+	for _, b := range blocks {
+		b := b
+
+		// Keep track of queried blocks.
+		resHints.AddQueriedBlock(b.meta.ULID)
+		indexr := b.indexReader()
+		cleanups = append(cleanups, func() {
+			// Defer all closes to the end of Series method.
+			runutil.CloseWithLogOnErr(s.logger, indexr, "series block")
+		})
+
+		// If query sharding is enabled we have to get the block-specific series hash cache
+		// which is used by blockSeries().
+		var blockSeriesHashCache *hashcache.BlockSeriesHashCache
+		if shardSelector != nil {
+			blockSeriesHashCache = s.seriesHashCache.GetBlockCache(b.meta.ULID.String())
+		}
+		g.Go(func() error {
+			var (
+				pstats *safeQueryStats
+				part   unloadedBatchSet
+				err    error
+			)
+
+			part, err = unloadedBucketBatches(
+				ctx, s.seriesPerBatch, indexr, b.meta.ULID, matchers, shardSelector, blockSeriesHashCache, chunksLimiter, seriesLimiter, req.SkipChunks, req.MinTime, req.MaxTime, req.Aggregates, s.logger)
+			if err != nil {
+				return errors.Wrapf(err, "fetch series for block %s", b.meta.ULID)
+			}
+
+			mtx.Lock()
+			batches = append(batches, part)
+			if pstats != nil {
+				stats = stats.merge(pstats.export())
+			}
+			mtx.Unlock()
+
+			return nil
+		})
+	}
+
+	s.mtx.RUnlock()
+	cleanup := func() {
+		for _, c := range cleanups {
+			c()
+		}
+	}
+
+	// Concurrently get data from all blocks.
+	{
+		begin := time.Now()
+		err := g.Wait()
+		if err != nil {
+			return nil, nil, nil, cleanup, err
+		}
+		stats.blocksQueried = len(batches)
+		stats.getAllDuration = time.Since(begin)
+	}
+
+	mergedBatches := mergedBatchSets(s.seriesPerBatch, batches...)
+	var set storepb.SeriesSet
+	if chunkReaders != nil {
+		set = newSeriesSetWithChunks(ctx, *chunkReaders, mergedBatches)
+	} else {
+		set = newSeriesSetWithoutChunks(mergedBatches)
+	}
+	return set, resHints, stats, cleanup, nil
+}
+
+type seriesSetWithoutChunks struct {
+	from unloadedBatchSet
+
+	currentIterator *unloadedBatchIterator
+}
+
+func newSeriesSetWithoutChunks(batches unloadedBatchSet) storepb.SeriesSet {
```

Comment:
We should preload in this case too, no? I mean preloading shouldn't just be about chunks but the symbolization of labels and fetching of chunks references too (which is what `bucketBatchSet.loadBatch()` does).